comment/tests/test_mixins.py

```diff
@@ -16,131 +21,175 @@ def test_non_ajax_request(self):
-        url_data.pop('app_name')
-        request = self.factory.get(self.get_url(**url_data))
-        response = self.mixin.dispatch(request)
+        self.url = reverse('comment:create')
+        self.client.force_login(self.user_1)
 
-        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
-        self.assertEqual(self.mixin.error, 'app name must be provided')
+    def test_logged_in_user_permission(self):
+        response = self.client.post(self.url, data=self.data, HTTP_X_REQUESTED_WITH='XMLHttpRequest')
+        self.assertEqual(response.status_code, 200)
 
-    def test_missing_model_type(self):
-        url_data = self.url_data.copy()
-        url_data.pop('model_name')
-        request = self.factory.get(self.get_url(**url_data))
-        response = self.mixin.dispatch(request)
+    @patch('comment.mixins.settings', COMMENT_ALLOW_ANONYMOUS=False)
+    def test_logged_out_user_permission(self, _):
+        self.client.logout()
+        response = self.client.post(self.url, data=self.data, HTTP_X_REQUESTED_WITH='XMLHttpRequest')
+        self.assertEqual(response.status_code, 302)
+        # user will be redirected to login
+        self.assertEqual(response.url, settings.LOGIN_URL + '?next=/comment/create/')
 
-        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
-        self.assertEqual(self.mixin.error, 'model name must be provided')
+    def test_permission_when_anonymous_comment_allowed(self):
+        self.assertTrue(settings.COMMENT_ALLOW_ANONYMOUS)
+        self.client.logout()
+        response = self.client.post(self.url, data=self.data, HTTP_X_REQUESTED_WITH='XMLHttpRequest')
+        self.assertEqual(response.status_code, 200)
 
-    def test_missing_model_id(self):
-        url_data = self.url_data.copy()
-        url_data.pop('model_id')
-        request = self.factory.get(self.get_url(**url_data))
-        response = self.mixin.dispatch(request)
 
-        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
-        self.assertEqual(self.mixin.error, 'model id must be provided')
+class ObjectLevelMixinTest(BaseCommentMixinTest):
+    def test_get_object_without_overriding(self):
+        object_mixin = ObjectLevelMixin()
+        self.assertRaises(ImproperlyConfigured, object_mixin.get_object)
 
-    def test_invalid_model_name(self):
-        url_data = self.url_data.copy()
-        model_name = 'not exists'
-        url_data['model_name'] = model_name
-        request = self.factory.get(self.get_url(**url_data))
-        response = self.mixin.dispatch(request)
 
-        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
-        self.assertEqual(self.mixin.error, f'{model_name} is NOT a valid model name')
+class CanEditMixinTest(BaseCommentMixinTest):
+    def setUp(self):
+        super().setUp()
+        self.client.force_login(self.user_1)
+        self.comment = self.create_comment(self.post_1, self.user_1)
```

Comment:
It would be better if we create objects in the `setUpTestData` method. It would make the tests run faster.